qiskit/transpiler/preset_passmanagers/common.py

```diff
@@ -0,0 +1,385 @@
+    """
+    if method == "unroller":
+        unroll = [Unroller(basis_gates)]
+    elif method == "translator":
+        unroll = [
+            # Use unitary synthesis for basis aware decomposition of
+            # UnitaryGates before custom unrolling
+            UnitarySynthesis(
+                basis_gates,
+                approximation_degree=approximation_degree,
+                coupling_map=coupling_map,
+                backend_props=backend_props,
+                plugin_config=unitary_synthesis_plugin_config,
+                method=unitary_synthesis_method,
+                target=target,
+            ),
+            UnrollCustomDefinitions(sel, basis_gates),
+            BasisTranslator(sel, basis_gates, target),
+        ]
+    elif method == "synthesis":
+        unroll = [
+            # # Use unitary synthesis for basis aware decomposition of
+            # UnitaryGates > 2q before collection
+            UnitarySynthesis(
+                basis_gates,
+                approximation_degree=approximation_degree,
+                coupling_map=coupling_map,
+                backend_props=backend_props,
+                plugin_config=unitary_synthesis_plugin_config,
+                method=unitary_synthesis_method,
+                min_qubits=3,
+                target=target,
+            ),
+            Unroll3qOrMore(target=target, basis_gates=basis_gates),
+            Collect2qBlocks(),
+            Collect1qRuns(),
+            ConsolidateBlocks(basis_gates=basis_gates, target=target),
+            UnitarySynthesis(
+                basis_gates=basis_gates,
+                approximation_degree=approximation_degree,
+                coupling_map=coupling_map,
+                backend_props=backend_props,
+                plugin_config=unitary_synthesis_plugin_config,
+                method=unitary_synthesis_method,
+                target=target,
+            ),
+        ]
+    else:
+        raise TranspilerError("Invalid translation method %s." % method)
+    return PassManager(unroll)
+
+
+def generate_scheduling(instruction_durations, scheduling_method, timing_constraints, inst_map):
+    """Generate a post optimization scheduling :class:`~qiskit.transpiler.PassManager`
+
+    Args:
+        instruction_durations (dict): The dictionary of instruction durations
+        scheduling_method (str): The scheduling method to use, can either be
+            ``'asap'``/``'as_soon_as_possible'`` or
+            ``'alap'``/``'as_late_as_possible'``
+        timing_constraints (TimingConstraints): Hardware time alignment restrictions.
+        inst_map (InstructionScheduleMap): Mapping object that maps gate to schedule.
+
+    Returns:
+        PassManager: The scheduling pass manager
+
+    Raises:
+        TranspilerError: If the ``scheduling_method`` kwarg is not a valid value
+    """
+    scheduling = PassManager()
+    if inst_map and inst_map.has_custom_gate():
+        scheduling.append(PulseGates(inst_map=inst_map))
```

Comment:
This seems to me more like optimization or translation. Pulse gate is indeed a `Schedule` but this is used to override the instruction set. Updating the instruction set doesn't perform any scheduling.

Reply:
I put it in scheduling since it was done after the optimization loop. I can make it a post-optimization stage or a pre-scheduling stage if you'd prefer that.

Reply:
Post-optimization sounds good to me. I still think we can have this in translation, because gate with calibration doesn't need to be translated or decomposed. For example, if I calibrate a CCX for (0, 1, 2), what I want to do would be
```python
circuit = QuantumCircuit(3)
circuit.ccx(0, 1, 2)
circuit.ccx(1, 2, 0)

backend.defaults().instruction_schdule_map.add("ccx", (0, 1, 2), my_toffoli_schedule)
transpiled = transpile(circuit, backend)
```
however, in this code, `ccx` is translated before reaching to the pulse gate pass. And no calibration is attached because no `ccx` primitive exist in the circuit. However, I cannot add `ccx` to basis gates because I have no calibration for ccx(1, 2, 0). This is why I want to move the pass before translation. Actually, we need to call pulse gate pass twice before and after the translation, because we may also want to add calibration for basis gates. Perhaps this can be solved with Target.